Import `docs` contents of all repositories

ChrisBAshton opened the pull request:
This PR pulls in top-level `*.md` files from the `docs/` folder of _all_ GOV.UK repositories, rather than just those opted in with `consume_docs_folder: true`. This is so that we surface all documentation more easily and have a consistent approach to pulling in external documentation. It also gives us the option of hosting documentation closer to the code (advantageous when documenting rake tasks, for example) and an opportunity to consolidate fragmented docs (deleting less useful docs and putting effort into maintaining one resource).

Trello: https://trello.com/c/UyfVvwYm/183-consume-all-docs-folders

## Visual changes

There are minimal visual changes. I've removed the https://docs.publishing.service.gov.uk/apis.html landing page and now external docs have app.html as a parent (as evidenced by the navigation at the top of the screenshot). There will be a follow-up PR (#2912) for improved visual changes.

|Before|After|
|------|------|
|<img width="1198" alt="Screenshot 2020-12-29 at 14 45 29" src="https://user-images.githubusercontent.com/5111927/103291954-bb610180-49e4-11eb-8ce8-59a46ad325a8.png">|<img width="1116" alt="Screenshot 2020-12-29 at 14 46 12" src="https://user-images.githubusercontent.com/5111927/103291956-bbf99800-49e4-11eb-8eaa-77ee4c2e7208.png">|

## Logic changes

- I've switched the [`consume_docs_folder` functionality](https://github.com/alphagov/govuk-developer-docs/pull/2595) to be opt out rather than opt in. That is, setting `consume_docs_folder: false` on an app in [applications.yml](https://github.com/alphagov/govuk-developer-docs/blob/59f31f9136812796fc39f46ad11b1a484ed1c03d/data/applications.yml#L203) will prevent the Developer Docs from pulling in content from its `docs` folder.
- I've removed the concept of 'API documentation'. There is now just 'App documentation', for all apps, regardless of whether or not the app is an API. Routes previously available at `/apis/**` will be redirected to their `/apps/**` equivalent. It is perhaps best explained with a diff, which demonstrates the more RESTful nature of these slugs:

```diff
  /apps/publishing-api.html
- /apis/publishing-api/dependency-resolution.html
+ /apps/publishing-api/dependency-resolution.html
 ```

I've also manually tested that private repositories' docs folders are not imported.

## Performance impact

- The compiled `search.json` is now 6.8 MB in size, compared to the current 5.7 MB, reflecting the extra documents that can be searched. The current one is "only" 1.3 MB when g-zipped, so there should only be a moderate dip in performance. The Middleman setup doesn't support GZip locally so it's hard to say for sure what the impact will be.
- As a result of the additional GitHub API requests we're making, the build time has increased from [about 2m26s](https://github.com/alphagov/govuk-developer-docs/runs/1599587458) to [about 3m15s](https://github.com/alphagov/govuk-developer-docs/pull/2910/checks?check_run_id=1605022404) (about a 25% increase). This is unfortunate but necessary, and I would say is still a reasonable build time.

benthorner reviewed (commented):
I've been looking forward to this - now we can see just how much (probably out-of-date 😬) documentation we actually have 😮 ☂️.

I've left a couple of more major comments where I think you've missed a few things. I also wondered: since none of the apps are opting out of consuming docs, can we just delete the code to support that? (It would make it quite a lot simpler.)

Finally, just a minor comment that I wasn't sure what the first commit was about, as it just repeats the diff. I ended up reading it a few times to try and understand it. For me, the important word to include would be "fix" e.g. "Fix returning no docs across all apps".

ChrisBAshton commented on source/templates/external_doc_template.html.erb:
```diff
@@ -1,6 +1,6 @@
 ---
 layout: api_layout
```
This is the only suggestion I'm hesitant about, as it involves a UI change. By removing the `api_layout`, we lose the menu that shows the imported docs, e.g.

<img width="321" alt="Screenshot 2020-12-29 at 16 56 42" src="https://user-images.githubusercontent.com/5111927/103300492-dc325280-49f6-11eb-98a1-4265e810f772.png">

I was hoping to tackle UI changes in #2912 instead. I experimented with moving these new menu items to the `app_layout` but it felt very cluttered, so this felt like a possible area of contention which I wanted to move to a subsequent PR rather than try to tack on here.

benthorner replied:
Oh, interesting. I see the `application_layout` is currently used [here](https://github.com/alphagov/govuk-developer-docs/blob/c3499bb0ee87677efbe6af38b47b511fb8edbb72/source/templates/application_template.html.md.erb#L2). However, I'm confused about how the layout affects the menu, which also lead me to question: how will this PR affect the current "Apps" menu (the one which shows "Application ownership by team")?

(I think there's definitely scope for improvement in the current navigation: it's really unclear what the relationship between the left sidebar and the header is - they seem to be totally different navigation systems!)

ChrisBAshton replied:
The layout affects the menu ('sidebar') [here](https://github.com/alphagov/govuk-developer-docs/blob/b081338b0a11cb561df200124e82e2533ed32728/source/layouts/application_layout.html.erb#L1-L24) - it sets the content for the sidebar. API layout [has its own sidebar](https://github.com/alphagov/govuk-developer-docs/blob/beade7407c23678a341374abcf1b35ab5d6c591d/source/layouts/api_layout.html.erb#L1-L16) which I've left relatively untouched for the purposes of just enabling 'consume docs' functionality across all repos. It therefore looks no different in this branch.

In a spike, I tried removing the API layout and moving the sidebar content into Application layout, but it gets very noisy. I'm taking a different approach in #2912.

benthorner replied:
Oh, I see. Thanks for explaining. And I see the home page sidebar is [statically inconsistent](https://github.com/alphagov/govuk-developer-docs/blob/b081338b0a11cb561df200124e82e2533ed32728/data/dashboard.yml) with the header links. Lovely! I think I see where you're going in the follow-up PR - I'll be interested to see what you do with the sidebar!

ChrisBAshton commented on app/proxy_pages.rb:
```diff
@@ -10,9 +10,7 @@ def self.resources
 
   def self.api_docs
```
`app_docs` already exists, so I've renamed the existing one to `app_overviews` (more appropriate, I reckon) and then applied the rename as suggested above.

benthorner replied:
Thanks - I agree that's clearer. Looking again, I see there's a terminology conflict in the use of "docs": sometimes it refers to "docs in the repo"; sometimes to "manual pages". For me, the most confusing instance is `AppDocs.apps_with_docs` 🤯. One other potential smell here is the encapsulation of `AppDocs`, which seems to be leaking into `ProxyPages` - the latter is taking on some of the responsibility of the former by adapting the GitHub response. Could making more use of the `App` class would address both these issues?

(This was already an issue, so out-of-scope for this PR. I think it would be good to follow-up on, though, as I imagine it will be confusing for someone looking at this code in future.)

ChrisBAshton replied:
`AppDocs.apps_with_docs` has never sat right with me! Agreed, we can refactor that out when we remove the `consume_docs_folder` functionality.

I seem to remember running into serialisation-related issues when delegating some of ProxyPage's concerns to the App, but will give this another go when that refactor PR comes. 👍

ChrisBAshton commented on data/applications.yml:
```diff
@@ -81,26 +81,22 @@
   team: "#govuk-platform-health"
   metrics_dashboard_url: https://grafana.blue.production.govuk.digital/dashboard/file/email_alert_api.json?refresh=10s&orgId=1
   production_hosted_on: aws
-  consume_docs_folder: true
```
Oops 🤦 probably kept this in to test something. Removed!

ChrisBAshton commented on app/github_repo_fetcher.rb:
```diff
@@ -29,8 +29,8 @@ def docs(app_name)
           path: "/apps/#{app_name}/#{filename}.html",
-          title: title,
-          markdown: contents,
+          title: title.to_s.force_encoding("UTF-8"),
```
Good idea. Done 👍

ChrisBAshton reviewed (commented):
Thanks @benthorner! RE removing `consume_docs_folder` functionality... it could certainly be removed, as no repos are currently using it. But I could see it being useful to be able to disable docs being imported per repo. I'd certainly want to keep it around for a little while in case when we roll this out we run into some unforeseen issues :) I can put it on my list to remove the code later if it hasn't been used in X weeks / months?

benthorner commented:
No worries - thanks for taking the time to look over my comments! Your replies have been helpful in giving me a bit more context about how this repo works. Just a couple of threads to tie up there, but hopefully nothing major.

What issues do you think might necessitate `consume_docs_folder: false`? I wonder if we could consider removing it sooner - it should be trivial to add the functionality back if we do need it in future (we'd need to do a PR anyway to update the config for whichever repo needed it). I agree we don't need to remove it in this PR, though 👍.

ChrisBAshton commented:
> What issues do you think might necessitate `consume_docs_folder: false`?

Perhaps a repo has too many noisy & not particularly useful documents, which are reducing the prominence of other search results. Or an imported document contains a special character that the import process crashes on (not that I know of any that would!). There may be no dev time to actually sit down and consolidate the documents, so it's quicker to just temporarily disable the docs import.

I agree we shouldn't need it for long though. Perhaps just a few days, once this is merged and we can prove it's all working fine on production.

I think I've addressed all your comments - let me know if I've missed anything! I suspect the more involved conversations may happen in #2912! 😄

benthorner commented on app/github_repo_fetcher.rb:
```diff
@@ -29,8 +29,8 @@ def docs(app_name)
-          title: title,
-          markdown: contents,
+          title: title.to_s.force_encoding("UTF-8"),
+          markdown: contents.to_s.force_encoding("UTF-8"),
```
Minor: I forgot to suggest this could also have a test to show the kind of thing it's protecting against.

ChrisBAshton replied:
The current test setup is [pretty horrific](https://github.com/alphagov/govuk-developer-docs/blob/master/spec/app/github_repo_fetcher_spec.rb#L84-L119) when mocking the actual file contents of a doc. Rather than add to the cruft, I'm hoping to refactor this in a future PR, where I'll retrospectively add a test for UTF-8 conversion 😅

benthorner replied:
Oh yeah, that's not very easy to extend 😬. Sounds like we're on the same page with all the follow-up changes 👍.

benthorner reviewed (approved):
> I agree we shouldn't need it for long though. Perhaps just a few days, once this is merged and we can prove it's all working fine on production.

Yup, that sounds reasonable to me.
